docs/Protected_build_porting_guide.md

```diff
@@ -0,0 +1,363 @@
+- User-space can access the kernel services through System Calls (downward communication)
+- Some information about user-space such as memory regions, and some user-space API's needs to be shared to kernel.
+- This information is stored in a special data structure called as the User Space Object which is linked into the kernel during build time.
+- The kernel can use the function pointers to user-space APIs available in the user space object to invoke required user space funtionality.
+
+
+## Steps for Porting Protected Build in a new chipset
+### Create a new configuration for Protected Build
+- Make a copy of the existing flat build configuration and rename it for protected build
+- For example: every chipset has a hello configuration under path build/config/CHIPSET_NAME.
+- Copy the contents of a config folder say "hello" under a new folder called “hello_protected”
+- Modify the defconfig and Make.defs as follows:
+
+#### Defconfig
+- The following configs need to be set or unset in the defconfig file.
+
+```
+CONFIG_BUILD_FLAT is not set
+CONFIG_BUILD_PROTECTED=y
+CONFIG_BUILD_2PASS=y
+CONFIG_PASS1_TARGET="all"
+CONFIG_PASS1_OBJECT=""
+CONFIG_TINYARA_USERSPACE=0x08080000
+CONFIG_ARM_HAVE_MPU_UNIFIED=y
+CONFIG_ARMV7M_MPU=y
+CONFIG_ARMV7M_MPU_NREGIONS=8
+CONFIG_MM_KERNEL_HEAP=y
+CONFIG_LIB_SYSCALL=y
+CONFIG_SYS_RESERVED=8
+CONFIG_SYS_NNEST=2
+```
+- TINYARA_USERPACE provides the address where the user-space blob is loaded into memory. TinyAra will expect to find an instance of "struct userspace_s' at this
+  location. The User Space Object is always placed at the start of the user memory region. This config value needs to be changed whenever there is a change in
+  the location where userspace blob is loaded and executed. This value must be set based on kernel-user split of Flash memory.
+
+
+#### Make.defs
+- The make.defs file provides several definitions which are used during the build process
+- One of the definitions is (ARCHSCRIPT) for the linker script to be used to link the kernel binary
+- In case of flat build, a single linker script is used - ld.script
+- In case of protected build this will be modified to provide two linker scripts
+	1. memory.ld : Provides the overall memory layout
+	2. kernel-space.ld : Provides the rules to link the kernel binary
+- The details of the two linker scripts is provided in the following sections.
+- The make.defs is modified as follows:
+
+```
+MEM_LDSCRIPT = memory.ld
+KSPACE_LDSCRIPT = kernel-space.ld
+....
+ARCHSCRIPT = -T$(TOPDIR)/../build/configs/$(CONFIG_ARCH_BOARD)/scripts/$(MEM_LDSCRIPT)
+ARCHSCRIPT += -T$(TOPDIR)/../build/configs/$(CONFIG_ARCH_BOARD)/scripts/$(KSPACE_LDSCRIPT)
+```
+
+
+### Creating Linker Scripts
+- In case of flat build, the single linker script ld.script describes the location and size of memory in the target as well as describes the memory layout of the output   file. In case of protected build we have three linker scripts to perform the above mentioned functionality:
```

Comment:
Please remove unnecessary space between output and file.

Reply:
Changes made